README.md
# kwikng 

[![NPM Version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]

A NodeJS Wrapper for [Kwik Delivery](https://kwik.delivery)

## Overview
This project provides an easy-to-use object-oriented API to access endpoints delineated at https://apikwik.docs.apiary.io/#reference OR https://liveapikwik.docs.apiary.io/#reference

## Getting Started

>Install from the NPM Registry

```bash

    $ npm i --save kwik-node

```

## API Values Format Codes ( Chart )

API REQUEST STATUSES:
====================

| Title               | Code | Description                |
|---------------------|------|----------------------------|
|PARAMETER_MISSING    | 100  |  A parameter is missing    |
|INVALID_KEY          | 101  |  Invalid access token      |
|ACTION_COMPLETE      | 200  |  Successful request        |
|ERROR_IN_EXECUTION   | 404  |  An Error Occured          |



TASK/JOB STATUSES:
=================

| Title     | Code | Description
|-----------|------|--------------------------------------------------------------------|
|UPCOMING   |  0   |  The task has been assigned to a agent                             |
|STARTED    |  1   |  The task has been started and the agent is on the way             |
|ENDED	    |  2   |  The task has been completed successfully                          |
|FAILED	    |  3   |  The task has not been completed successfully                      |
|ARRIVED    |  4   |  The task is being performed;agent has reached the destination     |
|UNASSIGNED |  6   |  The task has not been assigned to any agent                       |
|ACCEPTED   |  7   |  The task has been accepted by the agent which is assigned to him  |
|DECLINE    |  8   |  The task has been declined by the agent which is assigned to him  |
|CANCEL	    |  9   |  The task has been cancelled by the agent which is accepted by him |
|DELETED    |  10  |  The task is deleted by the agent                                  |



PAYMENT METHODS:
===============

| Title        | Code   | Description                                                |
|--------------|--------|------------------------------------------------------------|
|CARD          | 32     |  The task will be paid for using paystack card option       |
|CASH          | 8      |  The task will be paid for using cash in currency of local |
|STRIPE        | 2      |  The task will be paid for using stripe card option        |
|WALLET (PAGA) | 131072 |  The task will be paid for using a paga wallet option      |



VEHICLE SIZES:
=============

| Code | Description          |
|------|----------------------|
|  0   |  A motorcycle (bike) |
|  1   |  A small car         |
|  2   |  A medium car        |
|  3   |  A large truck       |



DELIVERY CHARGES:
================

| Code | Description                                                |
|------|------------------------------------------------------------|
|  1   |  Paid by the buyer (your own customer whom you deliver to) |
|  2   |  Paid by the kwik customer (you whom are using this API)   |



TIMEZONES:
=========

| Title                     | Code | Description                  |
|---------------------------|------|------------------------------|
|INDIAN STANDARD TIME (IST) | -330 |  offset from UTC in minutes  |
|CENTRAL AFRICAN TIME (CAT) | +180 |  offset from UTC in minutes  |
|WEST AFRICAN TIME (WAT)    | +60  |  offset from UTC in minutes  |


# Usage

```js

const Kwik = require('kwik-node')
const express = require('express')


const domainName = process.env.KWIK_DOMAIN_NAME; // 'app-test.kwik.delivery'
const environment = process.env.NODE_ENV; // 'production'

const kwikClient = new Kwik(
    domainName, 
    environment === 'production'
);

const app = express()
const port = 3000

/* setup a middleware to load up the API access_token via auth */
app.use('/create/delivery', async (req, res, next) => {
  if(req.method.toLowercase() === 'post'){
    let response = {
        body: {
            status:0,
            data: {
                access_token: '',
                vendor_details: {
                    vendor_id: '',
                    card_id: '',
                    user_id: ''
                }
            }
        }
    }

    try {
        response = await kwikClient.adminLogin({ 
            email: process.env.KWIK_ACCOUNT_EMAIL,
            password: process.env.KWIK_ACCOUNT_PASSWORD
        });
    } catch(error) {
        throw error;
    }


    kwikClient.setAccessToken(
        response.body.data.access_token
    );

    kwikClient.setVendorId(
        response.body.data.vendor_details.vendor_id
    );

    kwikClient.setUserId(
        response.body.data.vendor_details.user_id
    );

    kwikClient.setCardId(
        response.body.data.vendor_details.card_id
    );
  }

  next();
});



/* global error handler */
app.use(function (err, req, res, next) {
  res.status(500).send(err.message)
});


/* standard app endpoint to create a delivery task for dispatch by kwik dispatch riders */
app.post('/create/delivery', async (req, res) => {
    const today = new Date()
    const deliveries = [
            {
              address: req.body.delivery_address,
              name: req.body.business_name,
              latitude: req.body.choords.delivery_lat,
              longitude: req.body.choords.delivery_long,
              time: "2022-06-03 12:48:24",
              phone: req.body.mobile_number,
              email: req.body.email,
              has_return_task: false,
            }
    ];
    const pickups = [
            {
              address: req.body.pickup_address,
              name: "My Business name",
              latitude: req.body.choords.pickup_lat,
              longitude": req.body.choords.pickup_long,
              time: "2022-06-03 11:02:30",
              phone: "+2347045804049",
              email: "my.business@gmail.com"
            }
    ];

    /* NOTE: 
        no need to include 'domain_name', 'access_token' and 'vendor_id'
        as they are included automatically as long as the express middleware
        above obtains the details correctly in the middleware above using
        kwikClient.adminLogin() and sets all the details accordingly
    */

    /* HINT:
       It's vital to get Loaders and Vehicles info for your delivery task before
       that delivery task can be scheduled/created
    */

    /* Loaders info */
    const loaderDetails = await kwikClient.fetchAllLoadersOnAmount()

    /* Vehicles info */
    const vehicleDetails = await kwikClient.fetchAllAvailableDeliveryVehicles({
        size: 1 /* a small car (see API values format codes above 👆🏾) */
    });

    const deliveryTaskDetails = await kwikClient.getExactPricingForDeliveryTask({
        custom_field_template: 'pricing-template',
        auto_assignment: 1,
        layout_type: 1,
        pickup_custom_field_template: 'pricing-template',
        has_pickup: 1,
        has_delivery: 1,
        vehicle_id: vehicleDetails.body.data[0].vehicle_id,
        parcel_amount: 0,
        is_cod_job: 0, /* No collect on delivery */
        delivery_images: 'https://res.cloudinary.com/kwiky/image/upload/v1648282178/mova/item.png', /* The URL of the image of the item being delivered */
        delivery_instruction: 'Please, hand it over to the nanny', /* The instruction for the person who's to recieve the item being delivered */
        is_loader_required: loaderDetails.body.data.is_loader_enabled, /* equals 0; no loaders enabled */
        is_multiple_tasks: 1,
        payment_method: 32, /* payment via card */
        is_schedule_task: 0, /* task will not be scheduled in future */
        deliveries,
        pickups
    });

    console.log(
        'CURRENCY: ', 
        JSON.stringify(deliveryTaskDetails.body.data.currency, null, '\t')
    );

    console.log(
        'COST PER TASK (AMOUNT): ', 
        deliveryTaskDetails.body.data.per_task_cost
    );

    const cost_per_task = deliveryTaskDetails.body.data.per_task_cost

    const deliveryPaymentBreakDownDetails = await kwikClient.getEstimatedPriceForDeliveryTask({
        promo_value: 0,
        amount: String(cost_per_task),
        pickup_time: today.toISOString().replace('T', ' ').replace(/(?:\.[\d]{2,3}Z)/, ''),
        total_service_charge: deliveryTaskDetails.body.data.total_service_charge,
        vehicle_id: deliveryTaskDetails.body.data.vehicle_id,
        delivery_images: deliveryTaskDetails.body.data.delivery_images,
        delivery_instruction: deliveryTaskDetails.body.data.delivery_instruction,
        delivery_charge_by_buyer: 1, /* buyer pays for delivery charge (see API values format codes above 👆🏾) */
        is_cod_job: deliveryTaskDetails.body.data.is_cod_job,
        parcel_amount: deliveryTaskDetails.body.data.parcel_amount,
        is_loader_required: deliveryTaskDetails.body.data.is_loader_required,
    });

    const response = await kwikClient.scheduleDeliveryTask({
        team_id: 2, /* get your `team_id` from your admin dashboard */
        timezone: '+60', /* west african timezone (see API values format codes above 👆🏾) */
        auto_assignment: 1,
        layout_type: 1,
        is_multiple_tasks: 1,
        is_cod_job: deliveryTaskDetails.body.data.is_cod_job,
        delivery_charge_by_buyer: deliveryPaymentBreakDownDetails.body.data.delivery_charge_by_buyer,
        delivery_charge: deliveryPaymentBreakDownDetails.body.data.DELIVERY_CHARGE,
        collect_on_delivery: deliveryPaymentBreakDownDetails.body.data.COLLECT_ON_DELIVERY,
        surge_cost: deliveryPaymentBreakDownDetails.body.data.SURGE_PRICING,
        surge_type: deliveryPaymentBreakDownDetails.body.data.SURGE_TYPE,
        cash_handling_charges: deliveryPaymentBreakDownDetails.body.data.CASH_HANDLING_CHARGE,
        cash_handling_percentage: deliveryPaymentBreakDownDetails.body.data.CASH_HANDLING_PERCENTAGE,
        net_processed_amount: deliveryPaymentBreakDownDetails.body.data.NET_CASH_PROCEEDS,
        kwister_cash_handling_charge: deliveryPaymentBreakDownDetails.body.data.KWISTER_CASH_HANDLING_CHARGE,
        is_loader_required: deliveryTaskDetails.body.data.is_loader_required,
        delivery_images: deliveryTaskDetails.body.data.delivery_images,
        delivery_instruction: deliveryTaskDetails.body.data.delivery_instruction,
        insurance_amount: deliveryTaskDetails.body.data.insurance_amount,
        total_no_of_tasks: deliveryTaskDetails.body.data.total_no_of_tasks,
        total_service_charge: deliveryTaskDetails.body.data.total_service_charge,
        parcel_amount: deliveryTaskDetails.body.data.parcel_amount,
        payment_method: 32, /* payment via card (see API values format codes above 👆🏾) */
        amount: String(cost_per_task),
        deliveries: deliveryTaskDetails.body.data.deliveries,
        pickups: deliveryTaskDetails.body.data.pickups
    });

    res.status(response.body.status).send(response.body.message)
});

app.listen(port, () => {
    console.log(`Example app listening at http://localhost:${port}`)
});


```

## API Resources

>Each method expects an object literal with both **route parameters** and **request parameters (query / body)**. Please, go through the _src/Kwik/index.js_ `apiEndpoints` object to see the specific items that should make up the object literal for each method and their types

- **Tasks**
  - kwikClient.cancelDeliveryTask()
  - kwikClient.scheduleDeliveryTask()
- **Vehicles**
  - kwikClient.fetchAllAvailableDeliveryVehicles()
- **Loaders**
  - kwikClient.fetchAllLoadersOnAmount()
- **Pricing**
  - kwikClient.getExactPricingForDeliveryTask()
  - kwikClient.getEstimatedPriceForDeliveryTask()
- **Payments**
  - kwikClient.fetchAllMerchantCards()
  - kwikClient.addMerchantCard()
  - kwikClient.deleteMerchantCard()
- **Corporates**
  - kwikClient.createCorporate()
  - kwikClient.listAllCorporates()
  - kwikClient.listAllCorporatesInvoices()
  - kwikClient.listAllCustomersPerCorporates()
  - kwikClient.listAllCorporatesInvoicesSub()
- **Task Details ( Jobs )**
  - kwikClient.getSingleDeliveryTaskDetails()
  - kwikClient.getAllDeliveryTaskDetails()
  - kwikClient.getDeliveryTaskStatus()

# License

MIT

# Credits

- [Ifeora Okechukwu](https://twitter.com/isocroft)

# Contributing

See the [CONTRIBUTING.md](https://github.com/stitchng/kwikng/blob/master/CONTRIBUTING.md) file for info

[npm-image]: https://img.shields.io/npm/v/kwik-node.svg?style=flat-square
[npm-url]: https://npmjs.org/package/kwik-node

[travis-image]: https://img.shields.io/travis/stitchng/kwikng/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/stitchng/kwikng

## Support 

**Coolcodes** is a non-profit software foundation (collective) created by **Oparand** - parent company of StitchNG, Synergixe based in Abuja, Nigeria. You'll find an overview of all our work and supported open source projects on our [Facebook Page](https://www.facebook.com/coolcodes/).

>Follow us on facebook if you can to get the latest open source software/freeware news and infomation.

Does your business depend on our open projects? Reach out and support us on [Patreon](https://www.patreon.com/coolcodes/). All pledges will be dedicated to allocating workforce on maintenance and new awesome stuff.


CHANGELOG.md
<a name="0.0.8"></a>
# 0.0.8 (2022-03-27)

### Added Features
- Added the following enpoint wrappers: `listAllCustomersPerCorporates()`, `listAllCorporatesInvoicesSub()`, `getDeliveryTaskStatus()`, `fetchAllAvailableDeliveryVehicles()`, `fetchAllLoadersOnAmount()`

<a name="0.0.7"></a>
# 0.0.7 (2020-08-22)

### Bug Fixes
- Fixed issue logged by [@olaoluwa-98](https://github.com/olaoluwa-98) on [Issue 5](https://github.com/stitchng/kwikng/issues/5) with [this **PR**](https://github.com/stitchng/kwikng/pull/6)

<a name="0.0.6"></a>
# 0.0.6 (2020-07-18)

### Bug Fixes
- Fixed type error

<a name="0.0.5"></a>
# 0.0.5 (2020-05-25)

### Bug Fixes
- Fixed issue logged by [@olaoluwa-98](https://github.com/olaoluwa-98) on [Issue 3](https://github.com/stitchng/kwikng/issues/3) related to [this **got** issue](https://github.com/sindresorhus/got/issues/511)

<a name="0.0.2"></a>
# 0.0.2 (2020-05-13)

### Bug Fixes
- None




test/kwik-object.test.js
'use strict'

var chai = require('chai')
var expect = chai.expect
var should = chai.should()

describe('Kwik Instance Test(s)', function () {
  // Created Instance
  var Kwik = require('../src/Kwik/index.js')
  var instance = new Kwik('app.kwik.delivery', false)

  it('should have a function [mergeNewOptions]', function () {
    /* eslint-disable no-unused-expressions */
    expect((typeof instance.setAccessToken === 'function')).to.be.true
    expect((typeof instance.setVendorId === 'function')).to.be.true
    expect((typeof instance.setUserId === 'function')).to.be.true
    expect((typeof instance.setCardId === 'function')).to.be.true
    expect((typeof instance.getEstimatedPriceForDeliveryTask === 'function')).to.be.true
    expect((typeof instance.listAllCorporates === 'function')).to.be.true
    expect((typeof instance.createCorporate === 'function')).to.be.true
    expect((typeof instance.listAllCorporatesInvoices === 'function')).to.be.true
    expect((typeof instance.scheduleDeliveryTask === 'function')).to.be.true
    expect((typeof instance.cancelDeliveryTask === 'function')).to.be.true
    expect((typeof instance.getExactPricingForDeliveryTask === 'function')).to.be.true
    expect((typeof instance.getAllDeliveryTaskDetails === 'function')).to.be.true
    expect((typeof instance.getSingleDeliveryTaskDetails === 'function')).to.be.true
    /* eslint-enable no-unused-expressions */
  })

  it('should throw an error if method is called without required arguments', function () {
    try {
      instance.scheduleDeliveryTask()
    } catch (err) {
      should.exist(err)
    }
  })

  it('should throw an error if method is called with any arguments other than an object', function () {
    try {
      instance.addMerchantCard([])
    } catch (err) {
      should.exist(err)
    }
  })
})


src/Kwik/index.js
'use strict'

const got = require('got')
const querystring = require('querystring')
const _ = require('lodash')

const apiEndpoints = {
  adminLogin: {
    path: '/vendor_login',
    method: 'POST',
    send_json: true,
    params: { domain_name$: String, email: String, password: String, api_login: Number },
    param_defaults: { api_login: 1 },
    route_params: null
  },
  listAllCustomersPerCorporates: {
    path: '/get_all_customers_per_corporate',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, limit: Number, skip: Number },
    param_defaults: { limit: 10, skip: 0 },
    route_params: null
  },
  listAllCorporatesInvoicesSub: {
    path: '/list_all_job_invoice_sub',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, corporate_id$: Number, vendor_id$: Number, start_date: String, end_date: String, limit: Number, skip: Number },
    param_defaults: { limit: 10, skip: 0 },
    route_params: null
  },
  getDeliveryTaskStatus: {
    path: '/getJobStatus',
    method: 'GET',
    send_json: false,
    params: { unique_order_id$: String, customer_id: Number },
    param_defaults: null,
    route_params: null
  },
  fetchAllAvailableDeliveryVehicles: {
    path: '/getVehicle',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, is_vendor: Number, size: Number },
    param_defaults: { is_vendor: 1, size: 0 /* motorcycle (bike) */ },
    route_params: null
  },
  fetchAllLoadersOnAmount: {
    path: '/getLoaderList',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, is_vendor: Number },
    param_defaults: { is_vendor: 1 },
    route_params: null
  },
  fetchAllMerchantCards: {
    path: '/fetch_merchant_cards',
    method: 'POST',
    send_json: true,
    params: { access_token$: String, domain_name$: String, payment_method: Number, origin_booking: Number },
    param_defaults: { payment_method: 32 /* card payment */, origin_booking: 1 },
    route_params: null
  },
  addMerchantCard: {
    path: '/add_cards_view',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, domain_name$: String, client_email: String },
    route_params: null
  },
  deleteMerchantCard: {
    path: '/delet_merchant_card',
    method: 'POST',
    send_json: true,
    params: { access_token$: String, card_id$: String, payment_method: Number, domain_name$: String },
    param_defaults: { payment_method: 32 },
    route_params: null
  },
  createCorporate: {
    path: '/customer/add',
    method: 'POST',
    send_json: true,
    params: { address: String, domain: String, email: String, first_name: String, last_name: String, name: String, latitude: String, longitude: String, password: String, user_type: Number, form_ids: Number, customer_type: Number, company_rc_number: String, phone: String, sla_id: Number },
    param_defaults: { user_type: 1 },
    route_params: null
  },
  listAllCorporates: {
    path: '/list_all_corporates',
    method: 'GET',
    send_json: false,
    params: { access_token$: String },
    route_params: null
  },
  listAllCorporatesInvoices: {
    path: '/list_all_payment_invoice',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, corporate_id$: Number },
    route_params: null
  },
  getEstimatedPriceForDeliveryTask: {
    path: '/get_bill_breakdown',
    method: 'POST',
    send_json: true,
    params: { access_token$: String, benefit_type: Number, amount: String, insurance_amount: Number, domain_name$: String, total_no_of_tasks: Number, form_id: Number, user_id$: Number, promo_value: Number, credits: Number, total_service_charge: Number, vehicle_id: Number, delivery_images: String, delivery_instruction: String, delivery_charge_by_buyer: Number, is_cod_job: Number, parcel_amount: Number, is_loader_required: Number, loaders_amount: Number, loaders_count: Number },
    param_defaults: { insurance_amount: 0, total_no_of_tasks: 1, form_id: 2, benefit_type: 0, credits: 0, is_cod_job: 0, is_loader_required: 0, delivery_charge_by_buyer: 2 },
    route_params: null
  },
  cancelDeliveryTask: {
    path: '/cancel_vendor_task',
    method: 'POST',
    send_json: true,
    params: { access_token$: String, vendor_id$: Number, job_id$: String, job_status: Number, domain_name$: String },
    route_params: null,
    param_defaults: { job_status: 9 }
  },
  getExactPricingForDeliveryTask: {
    path: '/send_payment_for_task',
    method: 'POST',
    send_json: true,
    params: { vendor_id$: Number, custom_field_template: String, pickup_custom_field_template: String, form_id: Number, access_token$: String, is_multiple_tasks: Number, domain_name$: String, timezone: String, has_pickup: Number, has_delivery: Number, auto_assignment: Number, user_id$: Number, layout_type: Number, deliveries: Array, payment_method: Number, pickups: Array, is_cod_job: Number, parcel_amount: Number, is_loader_required: Number, vehicle_id: Number, loaders_amount: Number, loaders_count: Number, delivery_instruction: String, delivery_images: String, is_schedule_task: Number },
    param_defaults: { custom_field_template: 'pricing-template', pickup_custom_field_template: 'pricing-template', form_id: 2, payment_method: 131072 /* paga wallet payment */, is_multiple_tasks: 1, has_pickup: 1, has_delivery: 1, timezone: '+60' /* West African Time: +1:00hr from UTC */, auto_assignment: 0, layout_type: 0, is_cod_job: 0, is_loader_required: 0, is_schedule_task: 0 },
    route_params: null
  },
  getAllDeliveryTaskDetails: {
    path: '/get_order_history_with_pagination',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, limit: String, skip: String },
    param_defaults: { limit: 10, skip: 0 },
    route_params: null
  },
  getSingleDeliveryTaskDetails: {
    path: '/view_task_by_relationship_id',
    method: 'GET',
    send_json: false,
    params: { access_token$: String, unique_order_id: String },
    param_defaults: null,
    route_params: null
  },
  scheduleDeliveryTask: {
    path: '/create_task_via_vendor',
    method: 'POST',
    send_json: true,
    params: { pickup_delivery_relationship: Number, delivery_charge: Number, delivery_images: String, fleet_id: String, is_loader_required: Number, loaders_amount: Number, loaders_count: Number, cash_handling_charges: Number, cash_handling_percentage: Number, delivery_instruction: String, collect_on_delivery: Number, surge_cost: Number, is_cod_job: Number, surge_type: Number, pickups: Array, deliveries: Array, payment_method: Number, total_service_charge: Number, total_no_of_tasks: String, amount: String, insurance_amount: Number, vendor_id$: Number, access_token$: String, vehicle_id: Number, is_multiple_tasks: Number, domain_name$: String, timezone: String, has_pickup: Number, has_delivery: Number, auto_assignment: Number, team_id: Number, kwister_cash_handling_charge: String, net_processed_amount: Number, delivery_charge_by_buyer: Number, layout_type: Number },
    param_defaults: { insurance_amount: 0, fleet_id: '', pickup_delivery_relationship: 0, payment_method: 131072 /* paga wallet payment */, is_multiple_tasks: 1, is_loader_required: 0, has_pickup: 1, has_delivery: 1, timezone: '+60' /* West African Time: +1:00hr from UTC */, auto_assignment: 0, layout_type: 0, delivery_charge_by_buyer: 2, is_cod_job: 0 },
    route_params: null
  }
}

/*
 * Provides a convenience extension to _.isEmpty which allows for
 * determining an object as being empty based on either the default
 * implementation or by evaluating each property to undefined, in
 * which case the object is considered empty.
 */
_.mixin(function () {
  // reference the original implementation
  var _isEmpty = _.isEmpty
  return {
    // If defined is true, and value is an object, object is considered
    // to be empty if all properties are undefined, otherwise the default
    // implementation is invoked.
    isEmpty: function (value, defined) {
      if (defined && _.isObject(value)) {
        return !_.some(value, function (value, key) {
          return value !== undefined
        })
      }
      return _isEmpty(value)
    }
  }
}())

const isLiteralFalsey = (variable) => {
  return (variable === '' || variable === false || variable === 0)
}

const checkTypeName = (target, type) => {
  let typeName = ''
  if (isLiteralFalsey(target)) {
    typeName = (typeof target)
  } else {
    typeName = ('' + (target && target.constructor.name))
  }
  return !!(typeName.toLowerCase().indexOf(type) + 1)
}

const isTypeOf = (value, type) => {
  let result = false

  type = type || []

  if (typeof type === 'object') {
    if (typeof type.length !== 'number') {
      return result
    }

    let bitPiece = 0
    type = [].slice.call(type)

    type.forEach(_type => {
      if (typeof _type === 'function') {
        _type = (_type.name || _type.displayName).toLowerCase()
      }
      bitPiece |= (1 * (checkTypeName(value, _type)))
    })

    result = !!(bitPiece)
  } else {
    if (typeof type === 'function') {
      type = (type.name || type.displayName).toLowerCase()
    }

    result = checkTypeName(value, type)
  }

  return result
}

const _jsonify = (data) => {
  return isTypeOf(data, ["undefined", "null"]) || data === ""
    ? "null"
    : typeof data === "object"
    ? data instanceof Date
      ? data.toDateString()
      : "toJSON" in data
      ? data.toJSON().replace(/T|Z/g, " ")
      : JSON.stringify(data)
    : data;
}

const setInputValues = (config, inputs) => {
  let httpReqOptions = {}
  let inputValues = {}
  let label = ''

  switch (config.method) {
    case 'GET':
    case 'HEAD':
      label = 'query'
      break

    case 'POST':
    case 'PUT':
    case 'PATCH':
    case 'DELETE':
      label = 'body'
      break
  }

  httpReqOptions[label] = {}

  if (config.param_defaults) {
    inputs = Object.assign({}, config.param_defaults, inputs)
  }

  for (var input in config.params) {
    if (config.params.hasOwnProperty(input)) {
      let param = input.replace('$', '')
      let _input = inputs[param]
      let _type = config.params[input]
      let _required = false

      if ((input.indexOf('$') + 1) === (input.length)) {
        _required = true
      }

      if (_input === void 0 || _input === '' || _input === null) {
        if (_required) { throw new Error(`param: "${param}" is required but not provided; please provide as needed`) }
      } else {
        httpReqOptions[label][param] = isTypeOf(_input, _type)
          ? (label === 'query'
            ? querystring.escape(_jsonify(_input))
            : _jsonify(_input))
          : null

        if (httpReqOptions[label][param] === null) {
          throw new Error(`param: "${param}" is not of type ${_type.name}; please provided as needed`)
        }
      }
    } else {
      ; // @FIXME: Throw error maybe
    }
  }

  inputValues[label] = (label === 'body'
    ? (config.send_form
      ? httpReqOptions[label]
      : JSON.stringify(httpReqOptions[label])
    )
    : querystring.stringify(httpReqOptions[label]))

  return inputValues
}

const makeMethod = function (config) {
  let httpConfig = {
    headers: {
      'Cache-Control': 'no-cache',
      'Accept': 'application/json'
    },
    json: true
  }

  if (config.send_json) {
    httpConfig.headers['Content-Type'] = httpConfig.headers['Accept']
    httpConfig.form = false
  } else if (config.send_form) {
    httpConfig.headers['Content-Type'] = 'application/x-www-form-urlencoded'
    httpConfig.form = true
  }

  return function (requestParams = {}) {
    let pathname = false
    let payload = false

    if (!isTypeOf(requestParams, 'object')) {
      throw new TypeError('invalid argument type')
    }

    if (('domain_name$' in config.params)) {
      if (!requestParams.domain_name) {
        requestParams.domain_name = this.domainName
      }
    }

    if (('access_token$' in config.params)) {
      if (!requestParams.access_token) {
        requestParams.access_token = this.accessToken
      }
    }

    if (('vendor_id$' in config.params)) {
      if (!requestParams.vendor_id) {
        requestParams.vendor_id = this.vendorId
      }
    }

    if (('user_id$' in config.params)) {
      if (!requestParams.user_id) {
        requestParams.user_id = this.userId
      }
    }

    if (('card_id$' in config.params)) {
      if (!requestParams.card_id) {
        requestParams.card_id = this.cardId
      }
    }

    if (!_.isEmpty(requestParams, true)) {
      if (config.params !== null) {
        payload = setInputValues(config, requestParams)
      }

      pathname = config.path
    } else {
      if (config.params !== null ||
             config.route_params !== null) {
        throw new Error('requestParam(s) Are Not Meant To Be Empty!')
      }
    }

    if (payload === false) {
      payload = {}
    }

    for (let type in payload) {
      if (payload.hasOwnProperty(type)) {
        httpConfig[type] = (type === 'query') ? payload[type] : JSON.parse(payload[type])
      }
    }

    let reqVerb = config.method.toLowerCase()
    let baseUrl = this.httpClientBaseOptions.baseUrl

    return this.httpBaseClient[reqVerb](`${baseUrl}${pathname}`, httpConfig)
  }
}

class KwikAPI {
  constructor (domainName = 'app.kwik.delivery', isProd = true) {
    /* eslint-disable camelcase */
    var api_base = {
      sandbox: (domainName || '').startsWith('staging-') ? 'https://staging-api-test.kwik.delivery' : 'https://api.kwik.delivery',
      live: 'https://apicopy.kwik.delivery'
    }

    this.domainName = domainName

    this.accessToken = ''
    this.vendorId = ''
    this.userId = ''

    this.httpClientBaseOptions = {
      baseUrl: (!isProd ? api_base.sandbox : api_base.live)
    }

    /* eslint-enable camelcase */
    this.httpBaseClient = got
  }

  setAccessToken (token) {
    this.accessToken = token
  }

  setVendorId (id) {
    this.vendorId = id
  }

  setUserId (id) {
    this.userId = id
  }

  setCardId (id) {
    this.cardId = id
  }
}

for (let methodName in apiEndpoints) {
  if (apiEndpoints.hasOwnProperty(methodName)) {
    KwikAPI.prototype[methodName] = makeMethod(apiEndpoints[methodName])
  }
}

module.exports = KwikAPI
